### PR TITLE
improvements / bugfix in capped-regions partitioner

### PR DESCRIPTION
**Pull request by ryan-williams**
- surface alternate coverage-computation strategy, "exploding" reads into per-base, depth-1 tuples, controllable by a flag
  - opens a part of the performance-space where memory-use is not proportional to maximum depth of reads, partly inspired by discussions on #572 
- fix a bug with how read-starts are tallied given `loci` strings that start and stop in the interior of contigs
- rename the loci-partitioners

### CODE REVIEW

File: src/main/scala/org/hammerlab/guacamole/readsets/iterator/ContigCoverageIterator.scala [hunk 1]
```diff
@@ -53,11 +52,28 @@ case class ContigCoverageIterator(halfWindowSize: Int,
       else
         lastStart = start
 
-      ends.enqueue(end)
-      numAdded += 1
+      if (end + halfWindowSize > locus) {
```
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/readsets/iterator/ContigCoverageIterator.scala]: I think I follow this change but not 100%, could you add some comments here about what should be happening?

File: src/main/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDD.scala [hunk 2]
```diff
@@ -120,24 +114,55 @@ class CoverageRDD[R <: ReferenceRegion: ClassTag](@transient rdd: RDD[R])
...
+    // Compute the bounds of loci that this region should contribute 1 unit of coverage-depth to.
+    val lowerBound = math.max(0, start - halfWindowSize)
+    val upperBound = end + halfWindowSize
 
-      // Compute the bounds of loci that this region should contribute 1 unit of coverage-depth to.
-      contigLength = contigLengthsBroadcast.value(contigName)
-      lowerBound = math.max(0, start - halfWindowSize)
-      upperBound = math.min(contigLength, end + halfWindowSize)
+    // Iterator over loci spanned by the current region, including the half-window buffer on each end.
+    val readLocusIterator = new LociIterator(Iterator(Interval(lowerBound, upperBound)).buffered)
```
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDD.scala]: nit: mix of `region` and `read` here ... I don't love our use of `region` as our name for things with reference coordinates ... but until there is something better.

File: src/test/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDDSuite.scala [hunk 3]
```diff
@@ -26,88 +24,41 @@ class CoverageRDDSuiteRegistrar extends KryoTestRegistrar {
   }
 }
 
-class RegionRDDSuite
+class CoverageRDDSuite
   extends GuacFunSuite
     with RegionsRDDUtil
     with ContigLengthsUtil {
 
   override def registrar = "org.hammerlab.guacamole.readsets.rdd.CoverageRDDSuiteRegistrar"
```
Comment (arahuja) [inline src/test/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDDSuite.scala]: Doesn't matter much here since this class is defined in this file, but you can do `classOf[CoverageRDDSuiteRegistrar].getCanonicalName` to protect against re-namings or packages moving around.

File: src/main/scala/org/hammerlab/guacamole/readsets/iterator/ContigCoverageIterator.scala (hunk 1 above)
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/readsets/iterator/ContigCoverageIterator.scala]: yea, good idea, added a bunch!

also while explaining myself, found an edge case that I still wasn't handling correctly here, so fixed and added a test for it. thanks for prompting!

File: src/main/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDD.scala (hunk 2 above)
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDD.scala]: yea good point, done

File: src/test/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDDSuite.scala (hunk 3 above)
Comment (ryan-williams) [inline src/test/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDDSuite.scala]: great idea, changed here and elsewhere
